Complete Masterclass Refresh (#47)
balena-io/balena-cli-advanced-masterclass: README.md

# Balena CLI Advanced Masterclass

- **Masterclass Type:** Core
- **Maximum Expected Time To Complete:** 2 Hours

## Prerequisites

> To gain the most from this masterclass, we recommend that you first undertake the [balena CLI Masterclass](https://www.balena.io/docs/learn/more/masterclasses/cli-masterclass/)

Reviewer: ```suggestion
To gain the most from this masterclass, we recommend that you first go through the [balena CLI Masterclass](https://www.balena.io/docs/learn/more/masterclasses/cli-masterclass/)
```

## Introduction

> The balena Command Line Interface (balena CLI) consists of a number of commands that allow a user to develop, deploy, and manage balena fleets and devices.

Reviewer: ```suggestion
The balena Command Line Interface (balena CLI) consists of several commands that allow a user to develop, deploy, and manage balena fleets and devices.
```

> The balena Command Line Interface (balena CLI) utility consists of a number of commands that allow a user to develop, deploy, and manage balena fleets and devices.

Reviewer: ```suggestion
The balena Command Line Interface (balena CLI) consists of a number of commands that allow a user to develop, deploy, and manage balena fleets and devices.
```

Author: Good call, agreed!

> The previous balena CLI masterclass covered some of the most common techniques and topics. This masterclass aims to build on top of that, introducing you to additional features that can be used to gain finer control over the provisioning, deployment and management, of devices.

Reviewer: ```suggestion
The previous balena CLI masterclass covered some of the most common techniques and topics. This masterclass aims to build on top of that, introducing you to additional features that can be used to gain finer control over the provisioning, deployment, and management of devices.
```

In this masterclass, you will learn how to:

- Work with different balenaCloud environments
- Update balenaOS programmatically
- Preconfigure and preregister downloaded images
- Preload downloaded images with your application code

> If you have any questions about this masterclass as you proceed through it, or would like clarifications on any of the topics raised here, please do raise an issue on the repository this file is contained in, or contact us on the [balena forums](https://forums.balena.io/) where we'll be delighted to answer your questions.

Reviewer: Link the repo?

Author: Good call, will add

> The location of the repository that contains this masterclass and all associated code is: [https://github.com/balena-io/balena-cli-advanced-masterclass](https://github.com/balena-io/balena-cli-advanced-masterclass).

Reviewer: delete this line instead, and link the repo in the para above.

```

For the purposes of this document, we have chosen `'.'` (the current working directory) as the value for the `dataDirectory` property in the `balenarc.yml` file, but you could of course choose different directories such as `/home/ryanh/.balena/production` and `/home/ryanh/.balena/staging`. If this property is not specified, the default data directory is `.balena` or `_balena` (respectively on Linux/macOS and Windows) in the user's home directory.

> Watch out! The tilde character (`'~'`) is not resolved to the homedir when used inside a `balenarc.yml` file. Use the full directory such as `/home/ryanh/` instead of the tilde.

It is also worth noting that the CLI accepts the `BALENARC_DATA_DIRECTORY` environment variable as an alternative to the `dataDirectory` property in the `balenarc.yml` file. For example, the following command would allow the CLI to use the staging settings and token even if the current working directory was not `~/balenaStaging`:

```shell
$ cd /tmp
$ BALENARC_BALENA_URL=balena-staging.com \
  BALENARC_DATA_DIRECTORY=~/balenaStaging \
  balena whoami
== ACCOUNT INFORMATION
USERNAME: ryanh
EMAIL:    ryanh@balena.io

> URL:      balena-cloud.com

Reviewer: Is that the correct output? Above it the environment the command is targeting is the staging environment. 
Let me know, I could be wrong. 
```suggestion
URL:      balena-staging.com
```

Author: @vipulgupta2048 Good catch! This output shouldn't even be there I don't think.
